File: api/routes/screening.ts
/**
 * Screening route handler.
 * POST /api/screening/run
 * GET  /api/screening/candidate/:candidateId
 * GET  /api/screening/job/:jobId
 * GET  /api/screening/:id
 * Extracted verbatim from api/[...path].ts — lines 3323-3460.
 */
import type { VercelRequest, VercelResponse } from '@vercel/node';
import { getSupabaseAdmin } from '../_lib/supabase';
import { ok, badRequest, notFound, methodNotAllowed, requireAuth } from '../_lib/helpers';
import { generateJSON } from '../_lib/openai';
import { checkPlanAccess } from '../_lib/billing-utils';

export default async function handleScreening(req: VercelRequest, res: VercelResponse, segments: string[]) {
  const supabase = getSupabaseAdmin();

  // POST /api/screening/run
  if (segments.length === 2 && segments[1] === 'run') {
    const user = await requireAuth(req, res);
    if (!user) return;
    if (req.method !== 'POST') return methodNotAllowed(res);

    const billingGate = await checkPlanAccess(supabase, user.id, 'candidate_scoring');
    if (!billingGate.allowed) {
      return res.status(billingGate.status || 402).json(billingGate);
    }

    const { candidate_id, job_id } = req.body;
    if (!candidate_id || !job_id) return badRequest(res, `candidate_id (${candidate_id}) and job_id (${job_id}) are required`);

    const { data: candidate } = await supabase
      .from('candidates')
      .select('*')
      .eq('id', candidate_id)
      .single();

    if (!candidate) return notFound(res, `Candidate not found with ID: ${candidate_id}`);
    if (!candidate.resume_parsed_data) return badRequest(res, `Resume not parsed for candidate: ${candidate_id}. Please ensure the resume is uploaded and processed first.`);

    const { data: job } = await supabase
      .from('job_descriptions')
      .select('*')
      .eq('id', job_id)
      .eq('created_by', user.id)
      .single();

    if (!job) return notFound(res, 'Job not found');

    // Run ATS screening directly with AI
    const resumeJson = JSON.stringify(candidate.resume_parsed_data).slice(0, 6000);
    const prompt = `
Analyze this candidate's resume against the job requirements and provide ATS screening scores.

Job: ${job.title} (${job.role}, ${job.level})
Required Skills: ${(job.must_have_skills || []).join(', ')}
Nice-to-have Skills: ${(job.good_to_have_skills || []).join(', ')}
Min Experience: ${job.min_experience_years} years

Candidate Resume JSON:
${resumeJson}

Return JSON:
{
  "overall_score": 0-100,
  "skill_relevance_score": 0-100,
  "experience_score": 0-100,
  "education_score": 0-100,
  "credibility_score": 0-100,
  "shortlisted": true/false,
  "shortlist_reason": "...",
  "reason_codes": [{"code":"SKILL_MATCH","type":"positive","description":"...","impact":10}]
}`;

    let screeningResult: any;
    try {
      screeningResult = await generateJSON<any>(prompt, { timeoutMs: 20000, maxTokens: 1200, temperature: 0.1 });
    } catch (screenErr: any) {
      console.error('ATS screening AI failed:', screenErr?.message);
      return res.status(502).json({ error: 'Failed to run ATS screening. Please try again.' });
    }

    // Save screening result to database
    const dataToSave = {
      candidate_id: candidate_id,
      job_id: job_id,
      overall_score: screeningResult.overall_score,
      skill_relevance_score: screeningResult.skill_relevance_score ?? null,
      experience_score: screeningResult.experience_score ?? null,
      education_score: screeningResult.education_score ?? null,
      credibility_score: screeningResult.credibility_score ?? null,
      shortlisted: !!screeningResult.shortlisted,
      shortlist_reason: screeningResult.shortlist_reason ?? null,
      reason_codes: screeningResult.reason_codes ?? [],
      screened_at: new Date().toISOString(),
    };

    const { data: existing } = await supabase
      .from('ats_screenings')
      .select('id')
      .eq('candidate_id', candidate_id)
      .eq('job_id', job_id)
      .maybeSingle();

    const saved = existing
      ? await supabase
          .from('ats_screenings')
          .update(dataToSave)
          .eq('id', existing.id)
          .select('*')
          .maybeSingle()
      : await supabase
          .from('ats_screenings')
          .insert(dataToSave)
          .select('*')
          .maybeSingle();

    if (saved.error) return res.status(500).json({ error: saved.error.message });

    return ok(res, { success: true, screeningData: saved.data });
  }

  // GET /api/screening/candidate/:candidateId
  if (segments.length === 3 && segments[1] === 'candidate') {
    const user = await requireAuth(req, res);
    if (!user) return;
    if (req.method !== 'GET') return methodNotAllowed(res);

    const candidateId = segments[2];
    const { data, error } = await supabase
      .from('ats_screenings')
      .select('*')
      .eq('candidate_id', candidateId)
      .order('screened_at', { ascending: false });

    if (error) return res.status(500).json({ error: error.message });
    return ok(res, data);
  }

  // GET /api/screening/job/:jobId
  if (segments.length === 3 && segments[1] === 'job') {
    const user = await requireAuth(req, res);
    if (!user) return;
    if (req.method !== 'GET') return methodNotAllowed(res);

    const jobId = segments[2];
    const shortlistedOnly = req.query.shortlisted_only === 'true';
    const minScore = req.query.min_score ? parseInt(req.query.min_score as string) : null;

    let q = supabase.from('ats_screenings').select('*').eq('job_id', jobId);
    if (shortlistedOnly) q = q.eq('shortlisted', true);
    if (minScore !== null && !Number.isNaN(minScore)) q = q.gte('overall_score', minScore);

    const { data, error } = await q.order('overall_score', { ascending: false });
    if (error) return res.status(500).json({ error: error.message });
    return ok(res, data);
  }

  // GET /api/screening/:id
  if (segments.length === 2) {
    const user = await requireAuth(req, res);
    if (!user) return;
    if (req.method !== 'GET') return methodNotAllowed(res);

    const id = segments[1];
    const { data, error } = await supabase.from('ats_screenings').select('*').eq('id', id).single();
    if (error || !data) return notFound(res, 'Screening not found');
    return ok(res, data);
  }

  return notFound(res);
}

File: backend/app/api/v2/endpoints/screening.py
from __future__ import annotations

import json
import logging

from fastapi import APIRouter, Depends, Query

from app.api.v2.deps import require_user
from app.auth.clerk import ClerkUser
from app.services.ai.factory import get_ai
from app.services.db.supabase_service import get_db_admin_service
from app.utils.responses import ok, api_error
from app.services.ats_screening import get_ats_screening_service
from app.models.schemas import ResumeData, JobDescription

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/screening")


@router.post("/run")
async def run_screening(
    payload: dict,
    user: ClerkUser = Depends(require_user),
):
    """Node-compatible: POST /api/screening/run

    Body: { candidate_id, job_id }
    Returns: { success: true, screeningData: <row> }
    """

    candidate_id = payload.get("candidate_id")
    job_id = payload.get("job_id")
    if not candidate_id or not job_id:
        return api_error(message=f"candidate_id ({candidate_id}) and job_id ({job_id}) are required", status_code=400)

    db = get_db_admin_service()

    # Candidate
    candidates = await db.select("candidates", filters={"id": candidate_id}, limit=1)
    candidate = candidates[0] if candidates else None
    if not candidate:
        return api_error(message=f"Candidate not found with ID: {candidate_id}", status_code=404)

    if not candidate.get("resume_parsed_data"):
        return api_error(
            message=f"Resume not parsed for candidate: {candidate_id}. Please ensure the resume is uploaded and processed first.",
            status_code=400,
        )

    # Job (scope to created_by)
    jobs = await db.select("job_descriptions", filters={"id": job_id, "created_by": user.id}, limit=1)
    job = jobs[0] if jobs else None
    if not job:
        return api_error(message="Job not found", status_code=404)

    try:
        screening_service = get_ats_screening_service()
        
        resume_model = ResumeData(**(candidate.get("resume_parsed_data") or {}))
        job_model = JobDescription(
            id=str(job.get("id")),
            title=str(job.get("title")),
            role=str(job.get("role") or "custom"),
            level=job.get("level") or "mid",
            description=str(job.get("description") or ""),
            must_have_skills=job.get("must_have_skills") or [],
            good_to_have_skills=job.get("good_to_have_skills") or [],
            min_experience_years=job.get("min_experience_years") or 0,
        )
        
        result = await screening_service.screen_candidate(resume_model, "", job_model)
    except Exception as e:
        logger.error(f"Failed to screen candidate {candidate_id}: {e}", exc_info=True)
        return api_error(message="Failed to run ATS screening. Please try again.", status_code=502)

    data_to_save = {
        "candidate_id": candidate_id,
        "job_id": job_id,
        "overall_score": result.overall_score,
        "skill_relevance_score": result.skill_relevance_score,
        "experience_score": result.experience_score,
        "education_score": result.education_score,
        "credibility_score": result.credibility_score,
        "shortlisted": result.shortlisted,
        "shortlist_reason": result.shortlist_reason,
        "reason_codes": [rc.model_dump() for rc in result.reason_codes],
        "detailed_analysis": result.detailed_analysis.model_dump() if result.detailed_analysis else {},
    }

    # Upsert-like behavior (match Node)
    existing = await db.select("ats_screenings", filters={"candidate_id": candidate_id, "job_id": job_id}, limit=1)
    if existing:
        updated = await db.update("ats_screenings", data_to_save, filters={"id": existing[0]["id"]})
        saved_row = updated[0] if updated else None
    else:
        inserted = await db.insert("ats_screenings", [data_to_save])
        saved_row = inserted[0] if inserted else None

    if not saved_row:
        return api_error(message="Failed to save screening result", status_code=500)

    return ok({"success": True, "screeningData": saved_row})


@router.get("/candidate/{candidate_id}")
async def get_candidate_screenings(candidate_id: str, _user: ClerkUser = Depends(require_user)):
    """Node-compatible: GET /api/screening/candidate/:candidateId"""

    db = get_db_admin_service()

    def _fetch():
        return (
            db.client.from_("ats_screenings")
            .select("*")
            .eq("candidate_id", candidate_id)
            .order("screened_at", desc=True)
            .execute()
        )

    res = await db.run(_fetch)
    data = getattr(res, "data", None) or []
    return ok(data)


@router.get("/job/{job_id}")
async def get_job_screenings(
    job_id: str,
    shortlisted_only: bool = Query(default=False),
    min_score: int | None = Query(default=None, ge=0, le=100),
    _user: ClerkUser = Depends(require_user),
):
    """Node-compatible: GET /api/screening/job/:jobId"""

    db = get_db_admin_service()

    def _fetch():
        q = db.client.from_("ats_screenings").select("*").eq("job_id", job_id)
        if shortlisted_only:
            q = q.eq("shortlisted", True)
        if min_score is not None:
            q = q.gte("overall_score", int(min_score))
        return q.order("overall_score", desc=True).execute()

    res = await db.run(_fetch)
    data = getattr(res, "data", None) or []
    return ok(data)


@router.get("/{screening_id}")
async def get_screening(screening_id: str, _user: ClerkUser = Depends(require_user)):
    """Node-compatible: GET /api/screening/:id"""

    db = get_db_admin_service()

    def _fetch():
        return db.client.from_("ats_screenings").select("*").eq("id", screening_id).single().execute()

    res = await db.run(_fetch)
    data = getattr(res, "data", None)
    if not isinstance(data, dict):
        return api_error(message="Screening not found", status_code=404)
    return ok(data)
